src/basilisp/core.lpy

```diff
@@ -7496,6 +7407,108 @@
+
+(defmethod make-custom-data-readers basilisp.lang.interfaces/IPersistentMap
+  [mappings metadata]
+  (reduce (fn [m [k v]]
+            (let [v' (if (qualified-symbol? v)
+                       (intern (create-ns (symbol (namespace v)))
+                               (symbol (name v)))
+                       v)]
+              (cond
+                (not (qualified-symbol? k))
+                (throw
+                 (ex-info "Invalid tag in data-readers. Expected qualified symbol."
+                          (merge metadata {:form k})))
+
+                (not (ifn? v'))
+                (throw (ex-info "Invalid reader function in data-readers"
+                                (merge metadata {:form v})))
+
+                :else
+                (assoc m (with-meta k metadata) v'))))
+          mappings
+          mappings))
+
+(defmethod make-custom-data-readers importlib.metadata/EntryPoint
+  [entry-point metadata]
+  (make-custom-data-readers (.load entry-point)
+                            (assoc metadata
+                                   :basilisp.entry-point/name (.-name entry-point)
+                                   :basilisp.entry-point/group (.-group entry-point))))
+
+(defmethod make-custom-data-readers pathlib/Path
+  [file metadata]
+  (make-custom-data-readers
+   (with-open [rdr (basilisp.io/reader file)]
+     (read (if (.endswith (name file) "cljc")
+             {:eof nil :read-cond :allow}
+             {:eof nil})
+           rdr))
+   (assoc metadata :file (str file))))
+
+(defn- data-readers-entry-points []
+  (when (#{"true" "t" "1" "yes" "y"} (.lower
+                                      (os/getenv
+                                       "BASILISP_USE_DATA_READERS_ENTRY_POINT"
+                                       "true")))
+    (#?@(:lpy39-  [get (.entry_points importlib/metadata)]
+         :lpy310+ [.entry_points importlib/metadata ** :group])
+     "basilisp_data_readers")))
+
+(defn- data-readers-files
+  "Return a list of :external:py:class:`pathlib/Path`s pointing to
+  `data_readers.lpy` and `data_readers.cljc` found in each top
+  directory and immediate subdirectories of
+  the :external:py:data:`sys.path` entries. The list is ordered such
+  that entries with the filename `data_readers.lpy` appear first."
+  []
+  (->> sys/path
+       (mapcat (fn [dir] (when (os.path/isdir dir)
+                           (-> (comp
+                                (map #(os.path/join dir %))
+                                (mapcat (fn [sdir] (when (os.path/isdir sdir)
+                                                     (map #(os.path/join dir sdir %) (os/listdir sdir)))))
+                                (filter #(os.path/isfile %))
+                                (map pathlib/Path)
+                                (filter (comp #{"data_readers.lpy" "data_readers.cljc"} name)))
+                               (eduction (into ["."] (os/listdir dir)))))))
```

Comment:
```suggestion
  (->> sys/path
       (mapcat (fn [dir] (when (os.path/isdir dir)
                           (-> (comp
                                (mapcat (fn [^os/DirEntry sdir]
                                          (when (.is-dir sdir)
                                            (os/scandir (.-path sdir)))))
                                (filter #(.is-file %))
                                (map #(pathlib/Path (.-path %)))
                                (filter (comp #{"data_readers.lpy" "data_readers.cljc"} name)))
                               (eduction (os/scandir dir))))))
```

Since the original ticket mentions performance being a major concern, I attempted a quick rewrite using [`os.scandir`](https://docs.python.org/3/library/os.html#os.scandir) which Python's documentation claims can be a major speedup for cases when checking certain file attributes (which this code does repeatedly checking if a node is a file or directory which typically calls `stat` sys call and can be quite slow):

> Using [scandir()](https://docs.python.org/3/library/os.html#os.scandir) instead of [listdir()](https://docs.python.org/3/library/os.html#os.listdir) can significantly increase the performance of code that also needs file type or file attribute information, because [os.DirEntry](https://docs.python.org/3/library/os.html#os.DirEntry) objects expose this information if the operating system provides it when scanning a directory. All [os.DirEntry](https://docs.python.org/3/library/os.html#os.DirEntry) methods may perform a system call, but [is_dir()](https://docs.python.org/3/library/os.html#os.DirEntry.is_dir) and [is_file()](https://docs.python.org/3/library/os.html#os.DirEntry.is_file) usually only require a system call for symbolic links; [os.DirEntry.stat()](https://docs.python.org/3/library/os.html#os.DirEntry.stat) always requires a system call on Unix but only requires one for symbolic links on Windows.

I also removed `(into ["."] (os/listdir dir))` bits as it was causing issues with the expected `os.DirEntry` object (since it was just a string) but I also think it may have been unnecessary. Let me know if I'm mistaken.

Reply:
Excellent suggestion! performance is indeed paramount here.

The requirements where to look for data reader files are as follows, 
1. The root directory of each entry in `sys.path` (this is what the `"."` was triggering)
2. the immediate subdirectories of these entries (but no deeper).

I've updated your approach slightly to include the files from the root directories directly within `mapcat`.

Thanks

Reply:
Cool -- that looks good to me now. Thanks for the PR!